---
title: "Infer ASVS with DADA2"
author: "Alicia McElwee"
date: "`r Sys.Date()`"
output: html_document
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE,
                      fig.align = "center",
                      fig.path = "../figures/01_DADA2/") # Sends any figure output to this folder
```

# Before you start

## Set my seed
```{r set-seed}
# Setting the seed so randomized choices a reproducible
# Any number can be chosen
set.seed(5318008)

```

## Goals of this file

1. Use raw fastq files and generate quality plots to assess quality of reads.
2. Filter and trim out bad sequences and bases from our sequencing files.
3. Write out fastq files with high quality sequences
4. Evaluate the quality from our filter and trim.

# Installing and Loading Libraries

```{r install-libraries}

#install.packages("devtools")
library(devtools)

#devtools::install_github("benjjneb/dada2")
library(dada2)

```


# Load Data

```{r load-data}
# Set the raw fastq file path to the raw sequencing files
# Path to the fastq files
raw_fastqs_path <- "data/01_DADA2/01_raw_gzipped_fastqs"
raw_fastqs_path

# What files are in this path? Intuition Check
#list.files(raw_fastqs_path)

# How many files are there?
str(list.files(raw_fastqs_path))

# Create vector of forward reads
forward_reads <- list.files(raw_fastqs_path, pattern = "R1_001.fastq.gz", full.names = TRUE)
  # Intuition Check
head(forward_reads)

# Create vector of reverse reads
reverse_reads <- list.files(raw_fastqs_path, pattern = "R2_001.fastq.gz", full.names = TRUE)
  # Intuition Check
head(reverse_reads)
```


# Quality Plots
```{r raw-quality-plot}
# Randomly select 2 samples from dataset to evaluate
# random_samples <-
random_samples <- sample(1:length(reverse_reads), size = 2)

# Calculate and plot quality of these two samples
plotQualityProfile(forward_reads[random_samples])
plotQualityProfile(reverse_reads[random_samples])
```


# Prepare a placeholder for filtered reads
```{r prep-filtered-sequences}
# vector of our samples, extract sample name from files
# samples <- 
samples <- sapply(strsplit(basename(forward_reads), "_"), `[`,1)
head(samples)

# Place filtered reads into filtered_fastqs_path
filtered_fastqs_path <- "data/01_DADA2/02_filtered_fastqs"

# create 2 variables: filtered_F, filtered_R
filtered_forward_reads <- file.path(filtered_fastqs_path, paste0(samples,"_R1_filtered.fastq.gz"))
length(filtered_forward_reads)

filtered_reverse_reads <- file.path(filtered_fastqs_path, paste0(samples,"_R2_filtered.fastq.gz"))
length(filtered_reverse_reads)

```



# Filter and Trim Reads

Parameters of filter and trim **DEPENDS ON THE DATASET**

- `maxN` = number of N bases. Remove all Ns from the data for ASV                   creation
- `maxEE` = max expected errors, quality filtering threshold applied to expected errors. Here, it there's two expected errors its okay, but more than 2 throw away the sequence. Two values, first is for forward reads and second is for reverse reads. Better to do with just 1.

-`trimLeft` = the number of nucleotides to remove from the start of each read, here we are removing first 3 base pairs

-`trunQ` = removes any sequences that have base with quality less than 2

```{r filter-and-trim}
# NOTE: Must alter trim parameters to quality of your own dataset
# Assign a vector to filtered reads
# Trim out poor bases and first 3 bps on F reads
# write out filtered fastq files
filtered_reads <-
  filterAndTrim(fwd = forward_reads, filt = filtered_forward_reads,
              rev = reverse_reads, filt.rev = filtered_reverse_reads,
              maxN = 0, maxEE = c(2,2), trimLeft = 3,
              truncQ = 2, rm.phix = TRUE, compress = TRUE)
              #multithread = TRUE

```


# Running Python in R!
```{r setting-up-python}
# Use this code to set up python environment in R so can use {py} to make a python chunks
library(reticulate)
```


```{python}
no
version
```

# Session Information
```{r session-info}
# Ensure reproducibility by telling software versions
devtools::session_info()

```
