---
title: "Phyloseq PreProcessing"
author: "Alicia McElwee"
date: "`r Sys.Date()`"
output: 
  html_document:
    toc: yes
    toc_float: yes
    collapsed: no
    smooth_scroll: yes
    toc_depth: 3
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE,
                      fig.align = "center",
                      fig.path = "../figures/02_ProProcessing/") # Sends any figure output to this folder
```


# Load Libraries
```{r install-libraries}

pacman::p_load(devtools, phyloseq, tidyverse, dada2, install = FALSE)

```

# Goals

Here, we will process the data into a phyloseq object.

Three input files:

- ASV table
- Taxonomy table
- Track Reads (metadata)

Then, we will do the following:

1.) Remove Mitochondria
2.) Remove Chloroplasts
3.) Remove ASVs from negative controls and also negative control samples
4.) Evaluate Accuracy by looking at the Mock Community
5.) Remove samples without "enough" reads

Finally, write a data file of phyloseq output which will be stored in `data/02_PreProcessing/raw_preprocessed_phyloseq.RData`.


# Load Data

## ASV Table
```{r load-ASV-table}
# First, will load ASV table
load("data/01_DADA2/ASV_counts.RData")

# Inspect asv_tab
head(asv_tab)[,1:5]

# Fix sample names
sample_names <- colnames(asv_tab) # makes vector of sample filenames 
samples_fixed <- sapply(strsplit(basename(sample_names), "_"), `[`,1)
head(samples_fixed)

# re-write ASV count file to fix names
colnames(asv_tab) <- samples_fixed
str(asv_tab)

```

## Taxonomy Table
``` {r load-tax-table}
tax_df <- read.table("data/01_DADA2/ASV_taxonomy.tsv", sep = "\t", skip = 1)
head(tax_df)

# Fix column names
colnames(tax_df) <- c("asv_names","Kingdom","Phylum","Class", "Order", "Family", "Genus", "Species", "ASV", "ASV_seq")
head(tax_df)

# Make tax table a matrxi, Phyloseq needs tax table to be a matrix, requires row and column names
tax_mat <- tax_df %>%
  tibble::column_to_rownames(., var = "asv_names") %>% # takes asv_names column and makes it row names 
  as.matrix()


```

## Track Reads Data
``` {r load-track-reads}
load("data/01_DADA2/track_read_counts.RData")

# Take a look at the data
head(track_counts_df)
dim(track_counts_df)

#  Load in metadata
metadata_df <- read.csv("data/metadata.csv")
#view(metadata_df)
dim(metadata_df)
colnames(metadata_df)

# Merge metadata_df with track_reads_df
head(metadata_df)
head(track_counts_df) # Both share same names, so will use that to join
metadata_track_reads_df <- 
  metadata_df %>%
  left_join(., track_counts_df, by = "names") %>%
  dplyr::select(-X)

# Intuition Check
head(metadata_track_reads_df)

# Update row.names to be sample names
## Before
row.names(metadata_track_reads_df)

## Rewrite
row.names(metadata_track_reads_df) <- metadata_track_reads_df$names
  
## Check that it worked
row.names(metadata_track_reads_df)

# Intuition check
head(metadata_track_reads_df)
```


# Handoff to Phyloseq

Combine data into phyloseq object

```{r phyloseq-handoff}
# Double check its all good
dim(asv_tab)
dim(tax_mat)

# Intuition check
stopifnot(row.names(asv_tab) == row.names(tax_mat)) # Code stops if what is in parenthesis is not true
    # If no output, is good!

# Construct phyloseq object
raw_physeq <- phyloseq(otu_table(asv_tab, taxa_are_rows = TRUE), 
                       sample_data(metadata_track_reads_df),
                       tax_table(tax_mat))
raw_physeq

# Save this raw phyloseq object
save(raw_physeq, file = "data/02_PreProcessing/raw_physeq.RData")
```

# Clean up the data

Remove:

1. Chloroplast
2. Mitochondria

```{r rm-mitos-chloros}
# Remind myself of tax table
#view(tax_mat)

# Make new physeq without chloroplasts
noChloros_physeq <-
  raw_physeq %>%
  # rm chloroplasts
  subset_taxa(Order != "Chloroplast" | is.na(Order))

# How many taxa were chloroplasts?
num_chloro_ASVs <- ntaxa(raw_physeq) - ntaxa(noChloros_physeq)
num_chloro_ASVs

# Intuition check, checking that tax table of filtered data does not contain chloroplasts
#noChloros_physeq %>%
#  tax_table() %>%
#  data.frame() %>%
#  view()

# remove mitochondria
noChlorosMitos_physeq <- noChloros_physeq %>%
  subset_taxa(Family != "Mitochondria" | is.na(Family))

# How many mitochondrial ASVs?
num_mito_ASVs <-ntaxa(noChloros_physeq) - ntaxa(noChlorosMitos_physeq)
num_mito_ASVs

noChlorosMitos_physeq

# How many total ASVs were removed from Chloros and Mitos
total_ASVs_rm <- ntaxa(raw_physeq) - ntaxa(noChlorosMitos_physeq)
total_ASVs_rm

# Proportion ASVs kept?
ntaxa(noChlorosMitos_physeq)/ntaxa(raw_physeq)

```

# Evaluate and remove the control samples

Take a look at the negative controls and then make a decision about whether or not to remove the ASVS we found in our controls

1. Negative Controls
2. ASVs found in negative control and their distribution in the sample
3. Evaluate the mock community

## Negative Controls
```{r neg-controls}

# create a vector of samples that were negative controls
control_samples <- c("WaterControl", "022um-Control",
                     "3um-Control", "DNA-Ext-Control")

# Make a new phyloseq object of only the control samples
control_physeq <-
  noChlorosMitos_physeq %>%
  subset_samples(., (names %in% control_samples)) %>% # Found that controls had as many taxa as other samples
  #force remove ASVs with 0 counts
  prune_taxa(taxa_sums(.) > 0, .) # Get rid of 0 count ASVs in controls

control_physeq

# Are these 31 ASVs contaminates?

# Vector of control ASVs
control_ASVs <- 
  control_physeq %>%
  tax_table() %>%
  data.frame() %>%
  dplyr::select(ASV) %>%
  as.vector()

control_ASVs

# Visualize raw abundance values of control ASVs

control_physeq %>%
  plot_bar(., "ASV", fill = "ASV")
    # ASV 3 is most abundant, so 3rd most abundant ASV is in negative controls

# What is distribution of negative control ASVs between controls and actual samples?

######## Check the control ASVs in the sample
# Make new phyloseq object with only the controls ASVs in ALL samples
controlASV_physeq <- noChlorosMitos_physeq %>%
  subset_taxa(., ASV %in% control_ASVs$ASV)

# Inspect it
controlASV_physeq

# Take a look at the abundances
controlASV_physeq %>%
  plot_bar(., "ASV", fill = "ASV", facet_grid = Sample_or_Control~.)

  # Check ASVs that are in controls? If very abundant in controls and found in sample, likely want to remove them. Could do double check of taxonomy table of the ASVs to see what they may be/ are they likely to be contamination. Could also subtract out mean counts from controls

# Check tax of ASV_3
#control_physeq %>%
#  tax_table() %>%
#  data.frame() %>%
#  view()

# Would only like to keep ASV_18 and ASV_56, everything else is trash
controlASVs_toREMOVE <- dplyr::setdiff(control_ASVs$ASV, c("ASV_18","ASV_56"))
# double check
length(controlASVs_toREMOVE)

noControls_ChlorosMitos_physeq <-
  noChlorosMitos_physeq %>%
  # Actually remove the control ASVs (except 18, 56)
  subset_taxa(., !(ASV %in% controlASVs_toREMOVE)) %>%
  # REmove control samples
  subset_samples(., !(names %in% control_samples))
  
noControls_ChlorosMitos_physeq

# After filtering, how many ASVs were removed?
total_ASVs_rm <- ntaxa(raw_physeq) - ntaxa(noControls_ChlorosMitos_physeq)
total_ASVs_rm

```


# Evaluate Accuracy in Mock Community

- make a mock phyloseq object
- load in the standard fasta sequences of the mock community
- test for differences between the two

Looking to see how much of the mock community the sequencing was able to pull out.

```{r mock-community}
# Making phyloseq object of just mock community
mock_physeq <-
  noControls_ChlorosMitos_physeq %>%
  subset_samples(., names == "MockZymoPos") %>%
  prune_taxa(taxa_sums(.) > 0, .)

# Inspect it
mock_physeq

tax_table(mock_physeq)

# Write us a message summarizing this
cat("DADA2 Inferred", ntaxa(mock_physeq), "ASVs present in the mock community.")

##### Load in the mock reference sequence from Zymo
mock_ref <- getSequences("data/02_PreProcessing/mock_amplicons.fasta")
names(mock_ref) # To see names of the sequences

##### Pull our seqs from Sequenced Mock community
mock_seqs <-
  mock_physeq %>%
  tax_table() %>%
  data.frame() %>%
  dplyr::select(ASV, ASV_seq)
  
head(mock_seqs)
    
# WHich ASVs match reference?
matched_asvs <-
  mock_seqs %>%
  rowwise() %>%
  # check each ASV if they match, TRUE, if not == FALSE
  mutate(MATCH = any(grepl(ASV_seq,mock_ref))) %>%
  # Create vector of names that matched 
  pull(ASV)


# Evaluate which ones do NOT match??
# Whats their abudances?
# This helps provide error rate


cat(nrow(matched_asvs), "ASVs were exact matches to the expected reference sequence.")


# How many sequences were in the mock
mock_SeqDepth <-
  mock_physeq %>%
  otu_table() %>%
  data.frame() %>%
  colSums()


# Who is in the sequence mock community? How does it compare to what we expected?
mock_physeq %>%
  tax_table() %>%
  data.frame() %>%
  dplyr::select(Genus, Species)

# Compare to names in mock community
names(mock_ref)

# Curious: what are the mock counts in the actual samples?

noControls_ChlorosMitos_physeq %>%
  subset_taxa(., ASV %in% matched_asvs) %>%
  otu_table() %>%
  t() # Transpose so samples are rows and ASVs are columns

# Look at table to see if ASVs are highly present in the samples

```

What can we conclude about the mock community??

8 out of 8 ASV perfectly matched the sequences for the given ASVs. Although, it appears we dropped 1 sample during ASV creation since we know there are 9 sequences in the mock community. This appears to be one of the Salmonella species.

## Remove the mock community and its ASVs

```{r rm-mock-asvs}
length(matched_asvs)

# Make a new phyloseq object without mock and its asvs
noControlsMock_ChlorosMitos_physeq <- 
  noControls_ChlorosMitos_physeq %>%
  # remove mock community
    subset_samples(., names != "MockZymoPos") %>%
  # Remove ASVs from the mock
    subset_taxa(., !(ASV %in% matched_asvs))

noControlsMock_ChlorosMitos_physeq

# Intuition check: did it remove the 8 taxa
num_ASVs_mock <-
  ntaxa(noControls_ChlorosMitos_physeq) - 
  ntaxa(noControlsMock_ChlorosMitos_physeq)

stopifnot(num_ASVs_mock == 8)
  
```


# Evlauate the Sequencing Depth

```{r seq-depth}
# the Current data object
noControlsMock_ChlorosMitos_physeq

# Calculate the sum of sequences in each sample: what is the library size for each sample?
seqSums_df <- 
  noControlsMock_ChlorosMitos_physeq %>%
  otu_table() %>%
  # Sum each sample column
  colSums() %>%
  data.frame() %>%
  rownames_to_column(var = "names" ) %>%
  # joining sample depth onto other metadata, automatically removes samples not present in both
  left_join(., metadata_track_reads_df, by = "names") 

# Check
dim(seqSums_df)
head(seqSums_df)

# Rename second column
colnames(seqSums_df)[2] <- "TotalSeqs"

# Show the depth of samples
seqSums_df %>%
  dplyr::select(names, TotalSeqs) %>%
  arrange(TotalSeqs) 

# Seeing sample that has 0 reads


# plot it!
seqSums_df %>%
  ggplot(aes(x=reorder(names, TotalSeqs), y = TotalSeqs,
             fill = station)) +
  geom_bar(stat = "identity") 


# Density plot: tells overall density of sequence depths
seqSums_df %>%
  ggplot(aes(TotalSeqs, fill = station)) +
  geom_density(alpha = 0.5) 

# Ideally want the samples to have identical sequencing depths between different treatments

```

# Remove samples with few reads

```{r rm-samps-few-reads}
# Whats the min seq depth

min(sample_sums(noControlsMock_ChlorosMitos_physeq))

# Remove 20210615-MA-SCB2F
raw_preprocessed_physeq <-
  noControlsMock_ChlorosMitos_physeq %>%
  subset_samples(., names != "20210615-MA-SCB2F")

# Whats the new min seq depth
min(sample_sums(raw_preprocessed_physeq))

# Probably want to remove sample with less than 100 sequences?

```

# Save Preprocessed Phyloseq Object

```{r save-physeq}
save(raw_preprocessed_physeq,
     file = "data/02_PreProcessing/raw_preprocessed_physeq.RData")
```

# Session Information
```{r session-info}
# Ensure reproducibility by telling software versions
devtools::session_info()

```
