---
title: "Phylogenetic Tree Construction"
author: "Alicia McElwee"
date: "`r Sys.Date()`"
output: html_document
editor_options: 
  chunk_output_type: console
---


```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE,
                      fig.align = "center",
                      fig.path = "../figures/03_PhylogeneticTree/") # Sends any figure output to this folder
```

# Goals of this File

The goal is to create a phylogenetic tree!

1. Load in preprocessed phyloseq object
2. Create ASV fasta file from the phyloseq object
3. Align the 16S Sequences from fasta file with MAFFT.
4. Create a tree with FastTree2


## Load Packages & Phyloseq Object

```{r load-environment-data}
# phytool, gtree, RColorBrewer
pacman::p_load(tidyverse, phytools, ggtree, RColorBrewer, install = FALSE)

# Load Physeq
load("data/02_PreProcessing/raw_preprocessed_physeq.RData")
raw_preprocessed_physeq

set.seed(5318008)

```

# Create Fasta file of ASV and their Sequences

This fasta file will be used to create our alignment in MAFFT

```{r create-preprocessed-fasta}
# pull out ASV sequences and ASV names
asv_seq_df <-
  raw_preprocessed_physeq@tax_table %>%
  data.frame() %>%
  dplyr::select(ASV, ASV_seq)

# Add the > symbol to make fasta header
asv_seq_df$ASV <- paste0(">",asv_seq_df$ASV)
head(asv_seq_df)

# Create fasta object
asv_seq_fasta <- c(rbind(asv_seq_df$ASV, asv_seq_df$ASV_seq))
head(asv_seq_fasta)

# Write to a file
write(asv_seq_fasta, 
      file = "data/03a_PhylogeneticTree/preprocessed_ASVs.fasta")
```


# Align the 16S sequences from fasta file with MAFFT


```{r run-MAFFT, engine = 'bash', engine.opts = '-l'}
# TO test in shell directly from markdown, use Ctrl + Alt + Enter

# Write bash code to run MAFFT

export PATH=/programs/mafft/bin:$PATH

# Change directories to provide the fasta file we made above
cd data/03a_PhylogeneticTree/
pwd

# Set a seed
RANDOM=5318008

# Run Mafft
# For now, will use default options, note the version since default options change
# For own project,can use different parameters
# MAFFT automagically knows its a nucleotide alignment

/programs/mafft/bin/mafft --auto preprocessed_ASVs.fasta > mafft_aligned_ASVs.fasta

# Change back to project directory
cd ../../
pwd

```


# FastTree2

```{r run-FastTree, engine = 'bash', engine.opts = '-l'}

export PATH=/programs/FastTree-2.1.11:$PATH
FastTree
pwd

# cd into alignment folder
cd data/03a_PhylogeneticTree/
pwd

# Run FastTree to generate phylogenetic tree
#parameters:
  # -nt indicates its a nucleotide alignment
  # -gtr generalized time reversible model, nucleotide substitution model
  # -fastest speed up the model, reduce memory usage (recommended for datasets with more than 50,000 sequences)
  # -log output a log file
  # input alignment file
  # specify the output tree file
  
FastTree -nt -gtr -fastest -log FastTree.log mafft_aligned_ASVs.fasta > ASVs_unrooted.tree

cd ../../
echo "The working directory is $PWD"

```





# Session Information
```{r session-info}
# Ensure reproducibility by telling software versions
devtools::session_info()

```


