---
title: "Phylogenetic Tree Inspection and Rooting"
author: "Alicia McElwee"
date: "`r Sys.Date()`"
output: html_document
editor_options: 
  chunk_output_type: console
---

 # Set Environment
```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE,
                      fig.align = "center",
                      fig.path = "../figures/03_PhylogeneticTree/") # Sends any figure output to this folder
```

# Goals

1. Load fasttree unrooted tree
2. Add tree to our phyloseq object
3. Visualize and inspect tree with ggtree
4. Prune ASVs if needed
5. Root our tree
6. Combine new tree with a phyloseq object
7. Save 2 phyloseq objects: 1 with unrooted tree and 1 with a rooted tree

# Before you start
## Set Seed
```{r set-seed}

set.seed(5318008)

```


## Load Packages
```{r load-packages}
# phytool, gtree, RColorBrewer
pacman::p_load(tidyverse, phyloseq, phytools, ggtree, install = FALSE)

```

## Load Data
```{r load-data}
# Preprocessed phyloseq object
load("data/02_PreProcessing/raw_preprocessed_physeq.RData")
raw_preprocessed_physeq

# Load in the tree!
unrooted_tree <- read.tree(file= "data/03a_PhylogeneticTree/ASVs_unrooted.tree")
unrooted_tree
str(unrooted_tree)

```


# Merge Phyloseq and tree
```{r merge-physeq-and-tree}
# making sure tree has same number of taxa as phyloseq object
ntaxa(raw_preprocessed_physeq)
ntaxa(unrooted_tree)

unrooted_physeq <-
  merge_phyloseq(raw_preprocessed_physeq, unrooted_tree)
unrooted_physeq
```


# Plot tree with  `ggtree`

```{r plot-tree-unrooted}

# May be better to use another software such as FastTree or iTOL

# Make a basic tree
kingdom_tree <- 
  ggtree(unrooted_physeq) + 
  # color tips by kingdom 
  geom_tippoint(mapping = aes(color = Kingdom)) + 
  scale_color_manual(values = c("goldenrod1", "cornflowerblue", "grey")) +
  # Add title 
  labs(title = "Unrooted Tree") + 
  #move the legend to the bottom 
  theme(legend.position = "bottom"); kingdom_tree

# Make tree showing nodes
kingdom_node_tree <- ggtree(unrooted_physeq) +
  # color tips by kingdom
  geom_tippoint(mapping = aes(color = Kingdom)) +
  scale_color_manual(values = c("goldenrod1", "cornflowerblue", "grey"))+
  # Add title
  labs(title = "Unrooted Tree") +
  # Add information at the nodes, add node labels
  geom_text(aes(label=node), hjust = -0.5, vjust = -0.3, size = 2) +
  # move legend to the bottom
  theme(legend.position = "bottom"); kingdom_node_tree





```


# Evaluate the Long Branch 

This part is more manual
```{r eval-long-branch}
# View a specific clade
# Zoom in on original tree node 2537

viewClade(kingdom_node_tree +
          labs(title = "Unrooted Tree: Node 2537"),
          node =2537)

viewClade(kingdom_node_tree +
            labs(title = "Unrooted Tree: Node 2537") +
            geom_text(aes(label=ASV)),
          node = 2537)

# Long branch is ASV 456

# Lets make a note of the taxonomy and the ASV Sequence
unrooted_physeq %>%
  subset_taxa(., ASV == "ASV_456") %>%
  tax_table() %>%
  data.frame()

# Let's also check the counts of the ASV 
unrooted_physeq %>%
  subset_taxa(., ASV == "ASV_456") %>%
  otu_table() %>%
  data.frame() %>%
  colSums()




```

It’s an Alphaproteobacteria that has very few counts.

Let’s BLAST it on NSCBI Nucleotide Blast

It matches Protoeuglena noctilucae genome assembly, organelle: mitochondrion with an e-value of 2e-25 and a percent identity of 84.44%!!

This is definitely a contaminant that needs to be removed from the dataset. Let’s go ahead and do that…

# Prune ASV_456

```{r prune-ASV456}
# Function from Joey McMurdie: https://github.com/joey711/phyloseq/issues/652
pop_taxa = function(physeq, badTaxa){
  allTaxa <-  taxa_names(physeq)
  allTaxa <- allTaxa[!(allTaxa %in% badTaxa)]
  return(prune_taxa(allTaxa, physeq))}

# Lets use the pop_taxa function:
# Recreate a phyloseq object without ASV_456
unrooted_physeq_rm456 <-
  unrooted_physeq %>%
  pop_taxa(., "ASV_456")

# Intuition check
ntaxa(unrooted_physeq) - ntaxa(unrooted_physeq_rm456)

# Visually Inspect
ggtree(unrooted_physeq_rm456) +
  geom_tippoint(mapping = aes(color = Kingdom))


```

# Midroot our tree
```{r midroot-tree}
# Is the tree rooted?
unrooted_rm456_tree <- phy_tree(unrooted_physeq_rm456)
is.rooted(unrooted_rm456_tree)

# Midpoint root the tree
midroot_rm456_tree <- midpoint.root(unrooted_rm456_tree)

# Is the new tree rooted
is.rooted(midroot_rm456_tree)

# Assign to a new phyloseq object: merging subsetted phyloseq with the new rooted tree

# 1. Create the phyloseq object without ASV_456
physeq_rm456 <-
  raw_preprocessed_physeq %>%
  subset_taxa(ASV != "ASV_456")

# Merge tree with the new physeq_rm456
midroot_rm456_physeq <- merge_phyloseq(physeq_rm456, midroot_rm456_tree)
midroot_rm456_physeq

# Quick Inspection of tree
ggtree(midroot_rm456_physeq) +
  geom_tippoint(mapping = aes(color = Kingdom))

```

# Save to a new phyloseq object
```{r savep-physeq}
# Saving rooted and unrooted tree together
save(list = c("unrooted_physeq_rm456", "midroot_rm456_physeq"), 
     file = "data/03a_PhylogeneticTree/phytree_preprocessed_physeq.RData")

unrooted_physeq_rm456

midroot_rm456_physeq

```






# Session Information
```{r session-info}
# Ensure reproducibility by telling software versions
devtools::session_info()

```